### PR TITLE
ENH: grouped conditional coverage metrics

### LINKED ISSUE #309: Grouped Conditional Coverage Metrics
**Is your feature request related to a problem? Please describe.**
There is no metrics to compute conditional coverage that can estimate the model capacity to have good "local" coverage that depends on interval or set size (both classication and regression).

**Describe the solution you'd like**
Implement conditional coverage metrics for both classification and regression. 

**Describe alternatives you've considered**
To be more MAPIE friendly, it could be interesting to take account the type return of predict methods in the input of these new metrics.

**Additional context**
Angelopoulos, A. N., & Bates, S. (2021). A gentle introduction to conformal prediction and distribution-free uncertainty quantification. arXiv preprint arXiv:2107.07511.
Feldman, S., Bates, S., & Romano, Y. (2021). Improving conditional coverage via orthogonal quantile regression. Advances in Neural Information Processing Systems, 34, 2060-2071.

### PR DESCRIPTION
**Pull request by candicemyt**
# Description

There is no metrics to compute conditional coverage that can estimate the model capacity to have good "local" coverage that depends on interval or set size (both classification and regression).
Implement conditional coverage metrics for both classification and regression.

Fixes #309 

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test Regression SSC function on toyset
- [ ] Test Classification SSC function on toyset
- [ ] Test HSIC function on toyset

# Checklist

- [X] I have read the [contributing guidelines](https://github.com/simai-ml/MAPIE/blob/master/CONTRIBUTING.rst)
- [ ] I have updated the [HISTORY.rst](https://github.com/simai-ml/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/simai-ml/MAPIE/blob/master/AUTHORS.rst) files
- [X] Linting passes successfully : `make lint`
- [X] Typing passes successfully : `make type-check`
- [X] Unit tests pass successfully : `make tests`
- [X] Coverage is 100% : `make coverage`
- [ ] Documentation builds successfully : `make doc`

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -77,6 +77,7 @@ target/
 
 # VSCode
 .vscode
+.idea/
```
Comment (thibaultcordier) [inline .gitignore:80]: This is not the purpose of the PR, but I understand the usefulness of the line (no change requested).

File: mapie/metrics.py [hunk 2]
```diff
@@ -1,17 +1,18 @@
...
+                    check_array_shape_classification)
 
 
 def regression_coverage_score(
-    y_true: ArrayLike,
-    y_pred_low: ArrayLike,
-    y_pred_up: ArrayLike,
+        y_true: ArrayLike,
+        y_pred_low: ArrayLike,
+        y_pred_up: ArrayLike,
```
Comment (thibaultcordier) [inline mapie/metrics.py:15]: These changes modify the initial formatting (delete modification).

File: mapie/metrics.py [hunk 3]
```diff
@@ -53,8 +54,8 @@ def regression_coverage_score(
 
 
 def classification_coverage_score(
-    y_true: ArrayLike,
-    y_pred_set: ArrayLike
+        y_true: ArrayLike,
+        y_pred_set: ArrayLike
```
Comment (thibaultcordier) [inline mapie/metrics.py:58]: These changes modify the initial formatting (delete modification).

File: mapie/metrics.py [hunk 4]
```diff
@@ -103,8 +104,8 @@ def classification_coverage_score(
 
 
 def regression_mean_width_score(
-    y_pred_low: ArrayLike,
-    y_pred_up: ArrayLike
+        y_pred_low: ArrayLike,
+        y_pred_up: ArrayLike
```
Comment (thibaultcordier) [inline mapie/metrics.py:108]: These changes modify the initial formatting (delete modification).

File: mapie/metrics.py [hunk 5]
```diff
@@ -176,10 +177,10 @@ def classification_mean_width_score(y_pred_set: ArrayLike) -> float:
...
 
 def expected_calibration_error(
-    y_true: ArrayLike,
-    y_scores: ArrayLike,
-    num_bins: int = 50,
-    split_strategy: Optional[str] = None,
+        y_true: ArrayLike,
+        y_scores: ArrayLike,
+        num_bins: int = 50,
+        split_strategy: Optional[str] = None,
```
Comment (thibaultcordier) [inline mapie/metrics.py:183]: These changes modify the initial formatting (delete modification).

File: mapie/metrics.py [hunk 6]
```diff
@@ -230,12 +231,12 @@ def expected_calibration_error(
...
-    y_score_arg: Optional[ArrayLike] = None,
-    num_bins: int = 50,
-    split_strategy: Optional[str] = None,
-    classes: Optional[ArrayLike] = None,
+        y_true: ArrayLike,
+        y_scores: ArrayLike,
+        y_score_arg: Optional[ArrayLike] = None,
+        num_bins: int = 50,
+        split_strategy: Optional[str] = None,
+        classes: Optional[ArrayLike] = None,
```
Comment (thibaultcordier) [inline mapie/metrics.py:239]: These changes modify the initial formatting (delete modification).

File: mapie/metrics.py [hunk 7]
```diff
@@ -307,3 +308,397 @@ def top_label_ece(
         )
     ece /= len(labels)
     return ece
+
+
+def regression_coverage_score_v2(
+        y_true: NDArray,
+        y_intervals: NDArray,
```
Comment (thibaultcordier) [inline mapie/metrics.py:315]: Like previsously, use the initial formatting (delete tabs).

File: mapie/metrics.py [hunk 8]
```diff
@@ -307,3 +308,397 @@ def top_label_ece(
...
+            np.greater_equal(y_intervals[:, 1, :], y_true)
+        ),
+        axis=0
+    )
+    return coverages
+
+
+def classification_coverage_score_v2(
+        y_true: NDArray,
+        y_pred_set: NDArray
```
Comment (thibaultcordier) [inline mapie/metrics.py:348]: Like previsously, use the initial formatting (delete tabs).

File: mapie/metrics.py [hunk 9]
```diff
@@ -307,3 +308,397 @@ def top_label_ece(
...
+    coverage = np.take_along_axis(
+        y_pred_set, y_true, axis=1
+    ).mean(axis=0)
+    return coverage
+
+
+def regression_ssc(
+        y_true: NDArray,
+        y_intervals: NDArray,
+        num_bins: int = 3
```
Comment (thibaultcordier) [inline mapie/metrics.py:380]: Like previsously, use the initial formatting (delete tabs).

File: mapie/metrics.py [hunk 10]
```diff
@@ -307,3 +308,397 @@ def top_label_ece(
...
+        coverages[:, i] = regression_coverage_score_v2(y_true[indexes],
+                                                       intervals_binned)
+
+    return coverages
+
+
+def regression_ssc_score(
+        y_true: NDArray,
+        y_intervals: NDArray,
+        num_bins: int = 3
```
Comment (thibaultcordier) [inline mapie/metrics.py:445]: Like previsously, use the initial formatting (delete tabs).

File: mapie/metrics.py [hunk 11]
```diff
@@ -307,3 +308,397 @@ def top_label_ece(
...
+    >>> print(regression_ssc_score(y_true, y_intervals, num_bins=2))
+    [1.  0.5]
+    """
+    return np.min(regression_ssc(y_true, y_intervals, num_bins), axis=1)
+
+
+def classification_ssc(
+        y_true: NDArray,
+        y_pred_set: NDArray,
+        num_bins: Union[int, None] = None
```
Comment (thibaultcordier) [inline mapie/metrics.py:484]: Like previsously, use the initial formatting (delete tabs).

File: mapie/metrics.py [hunk 12]
```diff
@@ -307,3 +308,397 @@ def top_label_ece(
...
+                        axis=0
+                    )
+                )
+    return coverages
+
+
+def classification_ssc_score(
+        y_true: NDArray,
+        y_pred_set: NDArray,
+        num_bins: Union[int, None] = None
```
Comment (thibaultcordier) [inline mapie/metrics.py:573]: Like previsously, use the initial formatting (delete tabs).

File: mapie/metrics.py [hunk 13]
```diff
@@ -307,3 +308,397 @@ def top_label_ece(
...
+    ... ])
+    >>> print(classification_ssc_score(y_true,y_pred_set,num_bins=2))
+    [1.]
+    """
+    return np.min(classification_ssc(y_true, y_pred_set, num_bins), axis=1)
+
+
+def _gaussian_kernel(
+        x: NDArray,
+        kernel_size: int
```
Comment (thibaultcordier) [inline mapie/metrics.py:613]: Like previsously, use the initial formatting (delete tabs).

File: mapie/metrics.py [hunk 14]
```diff
@@ -307,3 +308,397 @@ def top_label_ece(
...
+    norm_x = x ** 2
+    dist = -2 * np.matmul(x, x.transpose((0, 2, 1))) \
+        + norm_x + norm_x.transpose((0, 2, 1))
+    return np.exp(-dist / kernel_size)
+
+
+def hsic(
+        y_true: NDArray,
+        y_intervals: NDArray,
+        kernel_sizes: ArrayLike = (1, 1)
```
Comment (thibaultcordier) [inline mapie/metrics.py:634]: Like previsously, use the initial formatting (delete tabs).

**Review by thibaultcordier — CHANGES_REQUESTED** [review]
Intermediate review: adapt the code with the initial formatting.

File: mapie/metrics.py [hunk 15]
```diff
@@ -307,3 +308,397 @@ def top_label_ece(
...
+) -> NDArray:
+    """
+    Effective coverage score obtained by the prediction intervals.
+
+    The effective coverage is obtained by estimating the fraction
+    of true labels that lie within the prediction intervals.
+
+    Parameters
+    ----------
+    y_true : NDArray of shape (n_samples,)
```
Comment (thibaultcordier) [inline mapie/metrics.py:325]: Spaces around a colon: no space before it and one space after it.

**Review by thibaultcordier — COMMENTED** [review]
Changes in all docstring.
Spaces around a colon: no space before it and one space after it.

File: mapie/metrics.py [hunk 16]
```diff
@@ -307,3 +308,397 @@ def top_label_ece(
...
+
+def regression_coverage_score_v2(
+    y_true: NDArray,
+    y_intervals: NDArray,
+) -> NDArray:
+    """
+    Effective coverage score obtained by the prediction intervals.
+
+    The effective coverage is obtained by estimating the fraction
+    of true labels that lie within the prediction intervals.
```
Comment (thibaultcordier) [inline mapie/metrics.py:321]: Add a brief comment to explain why it is different from the previous version (regression_coverage_score_v2 uses directly the output of predict method = intervals)

File: mapie/metrics.py [hunk 17]
```diff
@@ -307,3 +308,397 @@ def top_label_ece(
...
+
+def classification_coverage_score_v2(
+    y_true: NDArray,
+    y_pred_set: NDArray
+) -> NDArray:
+    """
+    Effective coverage score obtained by the prediction sets.
+
+    The effective coverage is obtained by estimating the fraction
+    of true labels that lie within the prediction sets.
```
Comment (thibaultcordier) [inline mapie/metrics.py:354]: Add a brief comment to explain why it is different from the previous version (classification_coverage_score_v2 uses directly the output of predict method = sets)

File: mapie/metrics.py [hunk 18]
```diff
@@ -307,3 +308,397 @@ def top_label_ece(
...
+    y_pred_set: NDArray,
+    num_bins: Union[int, None] = None
+) -> NDArray:
+    """
+    Compute Size-Stratified Coverage metrics proposed in [3] that is
+    the conditional coverage conditioned by the size of the predictions sets.
+    The sets are ranked by their size (ascending) and then divided into
+    num_bins groups : one value of coverage by groups is computed.
+
+
```
Comment (thibaultcordier) [inline mapie/metrics.py:492]: Remove new line.

**Review by thibaultcordier — COMMENTED** [review]
Request more details in function doctrings.

File: mapie/metrics.py [hunk 19]
```diff
@@ -307,3 +308,397 @@ def top_label_ece(
...
+    ... ])
+    >>> print(hsic(y_true, y_intervals))
+    [0.31787614 0.2962914 ]
+    """
+    y_true, y_intervals = check_array_shape_regression(y_true, y_intervals)
+    kernel_sizes = cast(NDArray, column_or_1d(kernel_sizes))
+    if len(kernel_sizes) != 2:
+        raise ValueError(
+            "kernel_sizes should be an ArrayLike of length 2"
+        )
```
Comment (thibaultcordier) [inline mapie/metrics.py:681]: Open question: is there a test that checks whether the ValueError is raised?

File: mapie/metrics.py [hunk 20]
```diff
@@ -307,3 +308,397 @@ def top_label_ece(
...
+    return coverages
+
+
+def classification_ssc_score(
+    y_true: NDArray,
+    y_pred_set: NDArray,
+    num_bins: Union[int, None] = None
+) -> NDArray:
+    """
+    Aggregate by the minimum for each alpha the Size-Stratified Coverage [3] :
```
Comment (thibaultcordier) [inline mapie/metrics.py:576]: Colon space to remove.

File: mapie/metrics.py [hunk 21]
```diff
@@ -307,3 +308,397 @@ def top_label_ece(
...
+    y_true, y_pred_set = check_array_shape_classification(y_true, y_pred_set)
+
+    sizes = np.sum(y_pred_set, axis=1)
+    if num_bins is not None:
+        check_number_bins(num_bins)
+        if num_bins > y_pred_set.shape[0]:
+            raise ValueError(
+                "Please provide a bin number lower \
+                than the number of observations"
+            )
```
[the remark this answers is not in the document]
Comment (candicemyt) [reply mapie/metrics.py:535]: Yes, there is one :  test_invalid_splits_classification_ssc(nums_bins=10)

File: mapie/utils.py [hunk 22]
```diff
@@ -1071,3 +1071,73 @@ def fix_number_of_classes(
...
+        raise ValueError(
+            f"shape mismatch between y_true {y_true.shape} \
+                and y_pred_set {y_pred_set.shape}"
+        )
+    if len(y_pred_set.shape) != 3:
+        if len(y_pred_set.shape) != 2:
+            raise ValueError(
+                "intervals should be a 3D array of shape \
+                (n_obs, n_classes, n_alpha)"
+            )
```
Comment (thibaultcordier) [inline mapie/utils.py:1105]: Open question: is there a test that checks whether the ValueError is raised?

File: mapie/utils.py [hunk 23]
```diff
@@ -1071,3 +1071,73 @@ def fix_number_of_classes(
...
+            raise ValueError(
+                "intervals should be a 3D array of shape (n_obs, 2, n_alpha)"
+            )
+        else:
+            y_intervals = np.expand_dims(y_intervals, axis=2)
+    if y_true.shape[0] != y_intervals.shape[0]:
+        raise ValueError(
+            f"shape mismatch between y_true {y_true.shape} \
+                and y_pred_set {y_intervals.shape}"
+        )
```
Comment (thibaultcordier) [inline mapie/utils.py:1142]: Open question: is there a test that checks whether the ValueError is raised?

**Review by thibaultcordier — COMMENTED** [review]
Open question to ensure that all tests have been thought through and additional format remarks

File: mapie/metrics.py (hunk 19 above)
Comment (candicemyt) [reply mapie/metrics.py:681]: Yes there is : test_hsic_invalid_kernel_sizes(kernel_sizes=[1])

File: mapie/metrics.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (thibaultcordier) [reply mapie/metrics.py:535]: Great :)

File: mapie/utils.py (hunk 22 above)
Comment (candicemyt) [reply mapie/utils.py:1105]: Yes there is : test_classification_y_pred_set_shape()
But "intervals" is an error it should be "y_pred_set" -> already fixed

File: mapie/utils.py (hunk 23 above)
Comment (candicemyt) [reply mapie/utils.py:1142]: Yes there is : test_regression_same_length()
Same as before : y_pred_set is an error, it should be "y_intervals" -> already fixed

File: mapie/tests/test_metrics.py [hunk 24]
```diff
@@ -219,3 +402,143 @@ def test_top_label_same_result() -> None:
...
+    np.testing.assert_allclose(
+        cond_cov_min, SSC_CLASSIF_COVERAGES_SCORE[params]
+    )
+
+
+@pytest.mark.parametrize("kernel_sizes", [[1], 2, [[1, 2]]])
+def test_hsic_invalid_kernel_sizes(kernel_sizes: ArrayLike):
+    """Test that invalid kernel sizes raises an error"""
+    with pytest.raises(ValueError):
+        hsic(y_toy, intervals, kernel_sizes=kernel_sizes)
```
Comment (thibaultcordier) [inline mapie/tests/test_metrics.py:525]: It may be useful to test negative values.

**Review by thibaultcordier — CHANGES_REQUESTED** [review]
I propose a test for the gaussian kernel

File: mapie/metrics.py (hunk 16 above)
Comment (thibaultcordier) [reply mapie/metrics.py:321]: Why is it resolved?

File: mapie/metrics.py (hunk 17 above)
Comment (thibaultcordier) [reply mapie/metrics.py:356]: Why is it resolved?

**Review by thibaultcordier — COMMENTED** [review]
Need distinction with existing previous functions.

File: mapie/tests/test_metrics.py [hunk 25]
```diff
@@ -5,23 +5,36 @@
...
                            classification_mean_width_score,
                            expected_calibration_error,
                            regression_coverage_score,
-                           regression_mean_width_score, top_label_ece)
+                           regression_mean_width_score, top_label_ece,
+                           regression_ssc,
+                           regression_ssc_score,
+                           classification_ssc,
+                           classification_ssc_score,
+                           hsic)
```
Comment (thibaultcordier) [inline mapie/tests/test_metrics.py:21]: Could you sort your import in alphabetical order?

File: mapie/metrics.py [hunk 26]
```diff
@@ -307,3 +311,407 @@ def top_label_ece(
...
+    ... [6.0, 9.0],
+    ... [9, 10.0]
+    ... ])
+    >>> print(regression_ssc(y_true, y_intervals, num_bins=2))
+    [[1. 1.]]
+    """
+    y_true, y_intervals = check_array_shape_regression(y_true, y_intervals)
+    check_number_bins(num_bins)
+    if num_bins > y_intervals.shape[0]:
+        raise ValueError(
```
Comment (vincentblot28) [inline mapie/metrics.py:433]: Add that an error might be raised in the docstring

File: mapie/metrics.py [hunk 27]
```diff
@@ -307,3 +311,407 @@ def top_label_ece(
...
+    >>> print(classification_ssc(y_true,y_pred_set,num_bins=2))
+    [[1. 1.]]
+    """
+    y_true, y_pred_set = check_array_shape_classification(y_true, y_pred_set)
+
+    sizes = np.sum(y_pred_set, axis=1)
+    if num_bins is not None:
+        check_number_bins(num_bins)
+        if num_bins > y_pred_set.shape[0]:
+            raise ValueError(
```
Comment (vincentblot28) [inline mapie/metrics.py:542]: Same comment than for regression

File: mapie/utils.py [hunk 28]
```diff
@@ -1071,3 +1071,73 @@ def fix_number_of_classes(
...
+    Parameters
+    ----------
+    y_true: ArrayLike
+        True labels.
+    y_pred_set: ArrayLike
+        Prediction sets given by booleans of labels.
+    """
+    y_true = cast(NDArray, column_or_1d(y_true))
+    if y_true.shape[0] != y_pred_set.shape[0]:
+        raise ValueError(
```
Comment (vincentblot28) [inline mapie/utils.py:1096]: add to docstring

File: mapie/utils.py [hunk 29]
```diff
@@ -1071,3 +1071,73 @@ def fix_number_of_classes(
...
+    y_true: NDArray
+        True labels.
+    y_intervals: NDArray
+        Lower and upper bound of prediction intervals
+        with different alpha risks.
+    """
+    y_true = cast(NDArray, column_or_1d(y_true))
+    if len(y_intervals.shape) != 3:
+        if len(y_intervals.shape) != 2:
+            raise ValueError(
```
Comment (vincentblot28) [inline mapie/utils.py:1133]: add to docstirng

File: mapie/metrics.py [hunk 30]
```diff
@@ -307,3 +311,407 @@ def top_label_ece(
...
+        Effective coverage obtained by the prediction intervals.
+    """
+    coverages = np.mean(
+        np.logical_and(
+            np.less_equal(y_intervals[:, 0, :], y_true),
+            np.greater_equal(y_intervals[:, 1, :], y_true)
+        ),
+        axis=0
+    )
+    return coverages
```
Comment (vincentblot28) [inline mapie/metrics.py:350]: very elegant !

File: mapie/metrics.py [hunk 31]
```diff
@@ -307,3 +311,407 @@ def top_label_ece(
...
+    y_pred_set: NDArray of shape (n_samples, n_class, n_alpha)
+        Prediction sets given by booleans of labels.
+
+    Returns
+    -------
+    NDArray of shape (n_alpha,)
+        Effective coverage obtained by the prediction sets.
+    """
+    if len(y_true.shape) < len(y_pred_set.shape):
+        y_true = np.expand_dims(y_true, axis=1)
```
Comment (vincentblot28) [inline mapie/metrics.py:380]: Isn't it always true ? because according ot your docstring, y_true is of shape (n_samples, ) and y_pred_set (n_samples, n_class, n_alpha)

File: mapie/metrics.py [hunk 32]
```diff
@@ -307,3 +311,407 @@ def top_label_ece(
         )
     ece /= len(labels)
     return ece
+
+
+def regression_coverage_score_v2(
```
Comment (vincentblot28) [inline mapie/metrics.py:316]: add a test to check that for the same value of alpha, regression_coverage_score_v2 and regression_coverage_score return the same value

File: mapie/metrics.py [hunk 33]
```diff
@@ -307,3 +311,407 @@ def top_label_ece(
...
+        np.logical_and(
+            np.less_equal(y_intervals[:, 0, :], y_true),
+            np.greater_equal(y_intervals[:, 1, :], y_true)
+        ),
+        axis=0
+    )
+    return coverages
+
+
+def classification_coverage_score_v2(
```
Comment (vincentblot28) [inline mapie/metrics.py:353]: add a test to check that for the same value of alpha, classification_coverage_score_v2 and classification_coverage_score  return the same value

File: mapie/metrics.py (hunk 31 above)
Comment (candicemyt) [reply mapie/metrics.py:380]: Yes, but in reality the function accepts y_pred_set of shape (n_samples, n_classes), should I change the docstring ?

Comment (thibaultcordier) [reply mapie/metrics.py:380]: It's important that the user calls the function correctly. I recommend a test that checks the shape of the input.

Comment (candicemyt) [reply mapie/metrics.py:380]: It's fixed, the function now requires y_true of shape (n_samples, n_alpha) and checks the shape of y_pred_set and y_true.

File: mapie/metrics.py [hunk 34]
```diff
@@ -307,3 +311,441 @@ def top_label_ece(
...
+def classification_ssc(
+    y_true: NDArray,
+    y_pred_set: NDArray,
+    num_bins: Union[int, None] = None
+) -> NDArray:
+    """
+    Compute Size-Stratified Coverage metrics proposed in [3] that is
+    the conditional coverage conditioned by the size of the predictions sets.
+    The sets are ranked by their size (ascending) and then divided into
+    num_bins groups : one value of coverage by groups is computed.
```
Comment (thibaultcordier) [inline mapie/metrics.py:525]: Space before the colon to remove.

File: mapie/metrics.py [hunk 35]
```diff
@@ -307,3 +311,441 @@ def top_label_ece(
...
+        True labels.
+    y_intervals: NDArray of shape (n_samples, 2, n_alpha) or (n_samples, 2)
+        Prediction sets given by booleans of labels.
+    kernel_sizes: int
+        The variance (sigma), this coefficient controls the width of the curve.
+
+    Returns
+    -------
+        NDArray of shape (n_alpha,)
+    One hsic correlation coefficient by alpha.
```
Comment (thibaultcordier) [inline mapie/metrics.py:698]: Problem with indentation (lines 697 and 698).

Comment (thibaultcordier) [reply mapie/metrics.py:698]: `NDArray of shape (n_alpha,)
    One hsic correlation coefficient by alpha.`

**Comment by thibaultcordier** [comment]
In my case, all I have to do is check the documentation (docstring) in readthedocs and I'm done.

File: mapie/metrics.py [hunk 36]
```diff
@@ -307,3 +311,441 @@ def top_label_ece(
...
+    -------
+    NDArray of shape (n_alpha,)
+        Effective coverage obtained by the prediction intervals.
+
+    Raises
+    ------
+    ValueError
+        If y_true is not of shape (n_samples, n_alpha)
+    """
+    y_intervals = check_array_shape_regression(y_true, y_intervals)
```
Comment (vincentblot28) [inline mapie/metrics.py:348]: Why do we ask the user to give y_true of shape (n_samples, n_alphas) ? the repeat of y_true should be done within the function

File: mapie/metrics.py [hunk 37]
```diff
@@ -307,3 +311,441 @@ def top_label_ece(
...
+    The effective coverage is obtained by estimating the fraction
+    of true labels that lie within the prediction sets.
+
+    It is different from ``classification_coverage_score`` because it uses
+    directly the output of ``predict`` method and can compute the
+    coverage for each alpha.
+
+    Parameters
+    ----------
+    y_true: NDArray of shape (n_samples, n_alpha)
```
Comment (vincentblot28) [inline mapie/metrics.py:379]: same comment for the shape of y_true

File: mapie/metrics.py [hunk 38]
```diff
@@ -307,3 +311,441 @@ def top_label_ece(
...
+            coverages[:, i] = classification_coverage_score_v2(
+                y_true[indexes],
+                np.take_along_axis(
+                    y_pred_set, np.expand_dims(indexes, axis=1), axis=0
+                )
+            )
+    else:
+        n_classes = y_pred_set.shape[1]
+        coverages = np.zeros((y_pred_set.shape[2], n_classes + 1))
+        for alpha in range(y_pred_set.shape[2]):
```
Comment (vincentblot28) [inline mapie/metrics.py:589]: why do you need to loop over the alphas here and not in your first "if" condition ?

File: mapie/metrics.py (hunk 36 above)
Comment (candicemyt) [reply mapie/metrics.py:348]: y_true does not have necessarily the same values for each alpha. For example, in regression_ssc, regression_coverage_score_v2 is called for each bin and the smallest intervals may not be the same for alpha1 and alpha2. In this case y_true[:, 0] and y_true[:, 1] are not the same.

File: mapie/metrics.py (hunk 38 above)
Comment (candicemyt) [reply mapie/metrics.py:589]: Because in the if condition the samples are splits with np.array_split() that returns a fixed number of sample per bin regardless of alpha. In the case num_bins=None, we create a split for each possible size of set : it is possible to have for alpha1 x sets of size 1 and y sets of size 1 for alpha2 (x!=y). In this case, the bins have not the same size for each alpha so we need a for loop.

File: mapie/metrics.py (hunk 36 above)
Comment (thibaultcordier) [reply mapie/metrics.py:348]: Do we talk about true targets or predicted targets?

Comment (candicemyt) [reply mapie/metrics.py:348]: True targets

File: mapie/tests/test_metrics.py [hunk 39]
```diff
@@ -33,7 +49,125 @@
         [False, True, False, True],
     ]
 )
-
+y_pred_set_2alphas = np.array([
```
Comment (vincentblot28) [inline mapie/tests/test_metrics.py:52]: I think this is a corner case because:
1. All possible sizes of prediction sets are represented
2. You have only one occurence of each size fr each lambda
3. You have the same number of occurence for each lambda
Could you trie with an example which have none of this particularities ?

File: mapie/metrics.py (hunk 36 above)
Comment (vincentblot28) [reply mapie/metrics.py:348]: I don't understand how the value of y_true can change with alpha

File: mapie/metrics.py (hunk 38 above)
Comment (vincentblot28) [reply mapie/metrics.py:589]: so it means that when we give a number of bins equal to 10 for instance, two observations with the same number of labels car belong to two different bins ?

File: mapie/metrics.py (hunk 36 above)
Comment (candicemyt) [reply mapie/metrics.py:343]: It can now take y_true of shapes (n_samples, n_alphas) and (n_samples,)

File: mapie/metrics.py (hunk 37 above)
Comment (candicemyt) [reply mapie/metrics.py]: same update

File: mapie/metrics.py (hunk 38 above)
Comment (candicemyt) [reply mapie/metrics.py]: I have deleted the if case, the two cases (num_bins=None and num_bins=int) are computed the same way. (It also solves the problem where two sets of the same size could have been in differents bins)

**Review by thibaultcordier — APPROVED** [review]
Comments submitted have been resolved.

File: mapie/metrics.py [hunk 40]
```diff
@@ -307,3 +313,426 @@ def top_label_ece(
...
+    >>> print(classification_ssc(y_true, y_pred_set, num_bins=2))
+    [[1.         0.66666667]]
+    """
+    y_true = cast(NDArray, column_or_1d(y_true))
+    y_pred_set = check_array_shape_classification(y_true, y_pred_set)
+
+    sizes = np.sum(y_pred_set, axis=1)
+    n_classes = y_pred_set.shape[1]
+    if num_bins is None:
+        bins = list(range(n_classes + 1))
```
Comment (vincentblot28) [inline mapie/metrics.py:560]: By doing so, in the case where you may have empty bins (if for instance their is no prediction sets with 5 labels). Is it possible to create the bins according the actual sizes that exists ?

File: mapie/metrics.py [hunk 41]
```diff
@@ -307,3 +313,426 @@ def top_label_ece(
...
+    [0.31787614 0.2962914 ]
+    """
+    y_true = cast(NDArray, column_or_1d(y_true))
+    y_intervals = check_array_shape_regression(y_true, y_intervals)
+    kernel_sizes = cast(NDArray, column_or_1d(kernel_sizes))
+    if len(kernel_sizes) != 2:
+        raise ValueError(
+            "kernel_sizes should be an ArrayLike of length 2"
+        )
+    if kernel_sizes[0] <= 0 or kernel_sizes[1] <= 0:
```
Comment (vincentblot28) [inline mapie/metrics.py:713]: (kernel_sizes <= 0).any()

File: mapie/metrics.py (hunk 40 above)
Comment (candicemyt) [reply mapie/metrics.py:560]: As it is actually implemented it returns nan if there is an empty bin. It is possible to change it and create the bins according the actual sizes that exists but the user will not know which bin is empty (which prediction set size does not exists) (it is not necessarily a problem)

File: mapie/metrics.py [hunk 42]
```diff
@@ -307,3 +313,426 @@ def top_label_ece(
...
+    Improving conditional coverage via orthogonal quantile regression.
+    Advances in Neural Information Processing Systems, 34, 2060-2071.
+
+    Parameters
+    ----------
+    y_true: NDArray of shape (n_samples,)
+        True labels.
+    y_intervals: NDArray of shape (n_samples, 2, n_alpha) or (n_samples, 2)
+        Prediction sets given by booleans of labels.
+    kernel_sizes: int
```
Comment (vincentblot28) [inline mapie/metrics.py:679]: it's not an int bu a Tuple[int]

File: mapie/metrics.py [hunk 43]
```diff
@@ -307,3 +313,426 @@ def top_label_ece(
...
+    [0.31787614 0.2962914 ]
+    """
+    y_true = cast(NDArray, column_or_1d(y_true))
+    y_intervals = check_array_shape_regression(y_true, y_intervals)
+    kernel_sizes = cast(NDArray, column_or_1d(kernel_sizes))
+    if len(kernel_sizes) != 2:
+        raise ValueError(
+            "kernel_sizes should be an ArrayLike of length 2"
+        )
+    if (kernel_sizes[0] <= 0).any():
```
Comment (vincentblot28) [inline mapie/metrics.py:713]: You must check for both values

Comment (candicemyt) [reply mapie/metrics.py:713]: Obviously ..

Comment (vincentblot28) [reply mapie/metrics.py:713]: and you should add this case to your test, as your tests are passing through this test

**Review by vincentblot28 — APPROVED** [review]

**Review by vincentblot28 — APPROVED** [review]
